docs/source/design/ui_design.rst

```diff
@@ -0,0 +1,328 @@
+######################
+User Interface Design
+######################
+
+This document outlines the design principles and guidelines for the user
+interface (UI) of NWChemEx. Our main goal is to provide a user-friendly and
+intuitive interface that enables users to carry out a variety of quantum
+chemistry calculations seamlessly. 
+
+
+Design Principles
+=================
+
+Performant
+-----------
+
+Performance is critical for NWChemEx with kernels supporting optimized MPI
+parallelization and GPU offloading. The UI is designed to uphold this
+performance standard and to provide the user with critical control elements to
+achieve high performance. This requires keeping the overhead of UI as minimal as
+possible by avoiding unnecessary copies or data movements and enabling user to
+access and modify the MPI communicator and GPU offloading as needed.
+ 
+Pythonic
```

Comment:
I agree that the note about not having DSLs and etc is important. I think I would also move this section above `Performant`. While performance is important, I think the "pythonic" aspect dictates a lot for the other sections.

Reply:
I used `Pythonic` since I thought it also captures `programmatic`. I have a discussion about parsers below in the minimal I/O part, but it is good to have it here, too. I agree on the DSL part and added back. I also moved this section to the top as @jwaldrop107 suggested. See 6230d5cfb57d167f90e77e4f71ddacda666046b2.